Geo refactor/handle agreement layer

yaguzmang opened the pull request:
Added stage management and UI components for handling the agreement layer. Additionally, added a custom hook to re-fetch the agreement layer when the level changes. The styles and form of the UI components are taken from the previous implementation. 


![Large GIF (846x762)](https://github.com/openforis/fra-platform/assets/41337901/1a6414f7-a828-4b9c-b2b6-669a6ce687d1)

yaguzmang committed:
Add LayerResponseData type and logic

Adding a response data type since the layers API not only
returns a mapId but also some other information, like palette,
which is necessary for the agreement layer.

yaguzmang committed:
Add logic to build the agreement layer req body

Adding a builder for the agreement layer calls, using the
section state with which the postLayer action was called. That
state is later used to get the selected layers and add them to
the layers array in the api call.

yaguzmang committed:
Add redux state management for the agreement layer

Adding logic to set the agreement level, and to set the palette
once the data is retrieved using the postLayer action.

yaguzmang committed:
Add custom hook to fetch agreement layer on level change

yaguzmang committed:
Add UI components for the agreement layer

Adding the UI components and logic for the agreement layer. The
styles are the same as before. I replaced some scss colors with
the defined generics we have, but not all colors were replaceable.
Further styling changes can be later added in a styling specific refactor.

minotogna commented on src/client/store/ui/geo/actions/postLayer.ts:
```diff
@@ -1,27 +1,27 @@
   'geo/post/layer',
-  async ({ countryIso, sectionKey, layerKey, layerState }) => {
-    const body = _getLayerRequestBody(countryIso, layerKey, layerState)
+  async ({ countryIso, sectionKey, layerKey, layerState, sectionState }) => {
```
Better not to pass the state . Here you have access to getState directly .

yaguzmang replied:
Good one 👍 Thanks!

minotogna commented on src/client/store/ui/geo/hooks/useAgreementLevel.ts:
```diff
@@ -0,0 +1,26 @@
+import { GeoActions } from '../slice'
+import { useGeoLayerSection } from '.'
+
+export const useAgreementLevel = (sectionKey: LayerSectionKey, layerKey: LayerKey) => {
```
This hooks doesn't belong to store. It dispatches an action . If it's used in one please , better to put in a hook folder where used. If common hook , under common hook folder

yaguzmang replied:
Okay, in that case, I'll move this one and useGteTreeCoverPercent (I based it off `useGetTableSections`).

yaguzmang replied:
Done, thanks!

minotogna commented on src/client/store/ui/geo/hooks/useAgreementLevel.ts:
```diff
@@ -0,0 +1,26 @@
+    if (prevAgreementLevel === agreementLevel) return // Skip if no change
+    if (agreementLevel === undefined) return // Skip when the property is not set
+    if (prevAgreementLevel === null) return // Skip with initial prev. value
+
```
Usually you can avoid getting the prev state , since a hook can run when a value changes . Check the logic .

yaguzmang replied:
Yes, after getting the state from the postLayer action, the usePrevious became unnecessary. Thanks!

yaguzmang committed:
Get LayerState and SectionState from within postLayer action

yaguzmang committed:
Update useAgreementLevel

Deleting the unnecessary hooks and updating after delegating the state
check to the postLayer action.

yaguzmang committed:
Move and rename useAgreementLevel to useFetchAgreementLevelLayer

Moving useAgreementLevel (now useFetchAgreementLevelLayer, as that is
more accurate and descriptive) to GeoMap/hooks.

yaguzmang committed:
Move and rename useGteTreeCoverPercent to useFetchGteTreeCoverPercentLayer

Moving this hook to GeoMap/hooks and giving it a better descriptive name as well.

yaguzmang committed:
Delete unnecessary usePrevious from useFetchGteTreeCoverPercentLayer

Since we now get the state from within postLayer, the usePrevious hook here
becomes unnecessary.

minotogna commented on src/client/pages/Geo/GeoMap/GeoMapMenuData/MapVisualizerPanel/LayersSectionPanel/components/AgreementLevelControl/AgreementLevelControl.tsx:
```diff
@@ -0,0 +1,128 @@
+
+                const style = level === currentSelectedLevel ? selectedStyle : genericStyle
+
+                return (
```
Exporting this as separate component would make the code easier to maintain

yaguzmang replied:
True. I added an `AgreementLevelSelector` component to the UI commit 👍

minotogna commented on src/client/pages/Geo/GeoMap/hooks/useFetchGteTreeCoverPercentLayer.ts:
```diff
@@ -0,0 +1,19 @@
+  useEffect(() => {
+    if (gteTreeCoverPercent === undefined) return // Skip when the property is not set
+    dispatch(GeoActions.postLayer({ countryIso, sectionKey, layerKey }))
+  }, [countryIso, layerKey, gteTreeCoverPercent, sectionKey, dispatch])
```
This is the same than useFetch agreement level ... can't it be one and only one fetch ? Then based on the layer , you parse the params ?

yaguzmang replied:
Wouldn't that make it harder to implement logic specific to those layers in the future? like caching (which would need to be handled differently for agreement).

yaguzmang replied:
Also, we need to add logic to re-fetch the agreement layer when the selected layers change. My initial thought was to do that from `toggleLayer` in the slice, but that is not possible since that would be a side effect. I suppose then it would need to go in this hook, making it different from the `useFetchGte...`. Any other suggestions of where that logic could go?
